File: src/main/java/nl/mpcjanssen/simpletask/FilterActivity.java
package nl.mpcjanssen.simpletask;


import android.app.ActionBar;
import android.app.ActionBar.Tab;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentStatePagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.design.widget.TabLayout;

import android.support.v7.app.AlertDialog;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.EditText;
import nl.mpcjanssen.simpletask.remote.FileStore;
import nl.mpcjanssen.simpletask.remote.FileStoreInterface;
import nl.mpcjanssen.simpletask.task.Priority;
import nl.mpcjanssen.simpletask.util.Util;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class FilterActivity extends ThemedActivity {

    final static String TAG = FilterActivity.class.getSimpleName();
    public static final String TAB_TYPE = "type";
    final static String CONTEXT_TAB = "context";
    final static String PROJECT_TAB = "project";
    final static String PRIO_TAB = "prio";
    final static String OTHER_TAB = "other";
    final static String SORT_TAB = "sort";
    final static String SCRIPT_TAB = "script";

    // Constants for saving state
    public static final String FILTER_ITEMS = "items";
    public static final String INITIAL_SELECTED_ITEMS = "initialSelectedItems";
    public static final String INITIAL_NOT = "initialNot";


    boolean asWidgetConfigure = false;
    ActiveFilter mFilter;

    TodoApplication m_app;
    SharedPreferences prefs;

    @Nullable
    private ViewPager pager;
    private Menu m_menu;
    private Logger log;
    private ScreenSlidePagerAdapter pagerAdapter;
    private FilterScriptFragment scriptFragment;

    @Override
    public void onBackPressed() {
        if (m_app.isBackSaving()) {
            applyFilter();
        }
        super.onBackPressed();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        log = LoggerFactory.getLogger(this.getClass());
        log.info("Called with intent: " + getIntent().toString());
        m_app = (TodoApplication) getApplication();
        prefs = TodoApplication.getPrefs();

        setContentView(R.layout.filter);

        Bundle arguments;

        Intent intent = getIntent();

        if (intent.getAction() != null) {
            asWidgetConfigure = getIntent().getAction().equals(AppWidgetManager.ACTION_APPWIDGET_CONFIGURE);
        }

        mFilter = new ActiveFilter();

        if (asWidgetConfigure) {
            mFilter.initFromPrefs(prefs);
        } else {
            mFilter.initFromIntent(intent);
        }


        pagerAdapter = new ScreenSlidePagerAdapter(getSupportFragmentManager());

        // Fill arguments for fragment
        arguments = new Bundle();
        arguments.putStringArrayList(FILTER_ITEMS,
                Util.sortWithPrefix(m_app.getTodoList().getContexts(), m_app.sortCaseSensitive(), "-"));
        arguments.putStringArrayList(INITIAL_SELECTED_ITEMS, mFilter.getContexts());
        arguments.putBoolean(INITIAL_NOT, mFilter.getContextsNot());
        arguments.putString(TAB_TYPE, CONTEXT_TAB);
        Fragment contextTab = new FilterListFragment();
        contextTab.setArguments(arguments);
        pagerAdapter.add(contextTab);


        // Fill arguments for fragment
        arguments = new Bundle();
        arguments.putStringArrayList(FILTER_ITEMS,
                Util.sortWithPrefix(m_app.getTodoList().getProjects(), m_app.sortCaseSensitive(), "-"));
        arguments.putStringArrayList(INITIAL_SELECTED_ITEMS, mFilter.getProjects());
        arguments.putBoolean(INITIAL_NOT, mFilter.getProjectsNot());
        arguments.putString(TAB_TYPE, PROJECT_TAB);
        Fragment projectTab = new FilterListFragment();
        projectTab.setArguments(arguments);
        pagerAdapter.add(projectTab);

        // Fill arguments for fragment
        arguments = new Bundle();
        arguments.putStringArrayList(FILTER_ITEMS, Priority.inCode(m_app.getTodoList().getPriorities()));
        arguments.putStringArrayList(INITIAL_SELECTED_ITEMS, Priority.inCode(mFilter.getPriorities()));
        arguments.putBoolean(INITIAL_NOT, mFilter.getPrioritiesNot());
        arguments.putString(TAB_TYPE, PRIO_TAB);
        Fragment prioTab = new FilterListFragment();
        prioTab.setArguments(arguments);
        pagerAdapter.add(prioTab);

        // Fill arguments for fragment
        arguments = new Bundle();
        arguments.putBoolean(ActiveFilter.INTENT_HIDE_COMPLETED_FILTER, mFilter.getHideCompleted());
        arguments.putBoolean(ActiveFilter.INTENT_HIDE_FUTURE_FILTER, mFilter.getHideFuture());
        arguments.putBoolean(ActiveFilter.INTENT_HIDE_LISTS_FILTER, mFilter.getHideLists());
        arguments.putBoolean(ActiveFilter.INTENT_HIDE_TAGS_FILTER, mFilter.getHideTags());
        arguments.putBoolean(ActiveFilter.INTENT_HIDE_CREATE_DATE_FILTER, mFilter.getHideCreateDate());
        arguments.putString(TAB_TYPE, OTHER_TAB);
        Fragment otherTab = new FilterOtherFragment();
        otherTab.setArguments(arguments);
        pagerAdapter.add(otherTab);


        // Fill arguments for fragment
        arguments = new Bundle();
        arguments.putStringArrayList(FILTER_ITEMS, mFilter.getSort(m_app.getDefaultSorts()));
        arguments.putString(TAB_TYPE, SORT_TAB);
        Fragment sortTab = new FilterSortFragment();
        sortTab.setArguments(arguments);
        pagerAdapter.add(sortTab);

        arguments = new Bundle();
        arguments.putString(ActiveFilter.INTENT_SCRIPT_FILTER, mFilter.getScript());
        arguments.putString(ActiveFilter.INTENT_SCRIPT_TEST_TASK_FILTER, mFilter.getScriptTestTask());
        arguments.putString(TAB_TYPE, SCRIPT_TAB);
        FilterScriptFragment scriptTab = new FilterScriptFragment();
        scriptFragment = scriptTab;
        scriptTab.setArguments(arguments);
        pagerAdapter.add(scriptTab);


        pager = (ViewPager)findViewById(R.id.pager);
        pager.setAdapter(pagerAdapter);
        // Give the TabLayout the ViewPager
        TabLayout tabLayout = (TabLayout) findViewById(R.id.sliding_tabs);
        tabLayout.setupWithViewPager(pager);
        tabLayout.setTabMode(TabLayout.MODE_SCROLLABLE);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.filter, menu);
        m_menu = menu;
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_filter_action:
                if (asWidgetConfigure) {
                    askWidgetName();
                } else {
                    applyFilter();
                }
                break;
            case R.id.menu_filter_load_script:
                openScript(new FileStoreInterface.FileReadListener() {
                    @Override
                    public void fileRead(final String contents) {
                        Util.runOnMainThread(
                                new Runnable() {
                                    @Override
                                    public void run() {
                                        setScript(contents);
                                    }
                                }
                        );
                    }
                });
                break;
        }
        return true;
    }

    private void openScript(final FileStoreInterface.FileReadListener file_read) {
        Util.runOnMainThread(new Runnable() {
            @Override
            public void run() {
                FileStore.FileDialog dialog = new FileStore.FileDialog(FilterActivity.this, new File(m_app.getTodoFileName()).getParent(), false);
                dialog.addFileListener(new FileStoreInterface.FileSelectedListener() {
                    @Override
                    public void fileSelected(final String file) {
                        new Thread(new Runnable() {

                            @Override
                            public void run() {
                                try {

                                    m_app.getFileStore().readFile(file, file_read);
                                } catch (IOException e) {
                                    Util.showToastShort(FilterActivity.this, "Failed to load script.");
                                    e.printStackTrace();
                                }
                            }
                        }).start();
                    }
                });
                dialog.createFileDialog(FilterActivity.this, m_app.getFileStore());
            }
        });
    }

    @NonNull
    private Intent createFilterIntent() {
        Intent target = new Intent(this, Simpletask.class);
        target.setAction(Constants.INTENT_START_FILTER);
        updateFilterFromFragments();
        mFilter.setName(mFilter.getProposedName());
        mFilter.saveInIntent(target);

        target.putExtra("name", mFilter.getProposedName());
        return target;
    }

    private void updateFilterFromFragments() {
        ArrayList<String> items;
        for (Fragment f : pagerAdapter.getFragments()) {
            switch (f.getArguments().getString(TAB_TYPE,"")) {
                case "":
                    break;
                case OTHER_TAB:
                    FilterOtherFragment of = (FilterOtherFragment) f;
                    mFilter.setHideCompleted(of.getHideCompleted());
                    mFilter.setHideFuture(of.getHideFuture());
                    mFilter.setHideLists(of.getHideLists());
                    mFilter.setHideTags(of.getHideTags());
                    mFilter.setHideCreateDate(of.getHideCreateDate());
                    break;
                case CONTEXT_TAB:
                    FilterListFragment lf = (FilterListFragment) f;
                    mFilter.setContexts(lf.getSelectedItems());
                    mFilter.setContextsNot(lf.getNot());
                    break;
                case PROJECT_TAB:
                    FilterListFragment pf = (FilterListFragment) f;
                    mFilter.setProjects(pf.getSelectedItems());
                    mFilter.setProjectsNot(pf.getNot());
                    break;
                case PRIO_TAB:
                    FilterListFragment prf = (FilterListFragment) f;
                    mFilter.setPriorities(prf.getSelectedItems());
                    mFilter.setPrioritiesNot(prf.getNot());
                    break;
                case SORT_TAB:
                    FilterSortFragment sf = (FilterSortFragment) f;
                    mFilter.setSort(sf.getSelectedItem());
                    break;
                case SCRIPT_TAB:
                    FilterScriptFragment scrf = (FilterScriptFragment) f;
                    mFilter.setScript(scrf.getScript());
                    mFilter.setScriptTestTask(scrf.getTestTask());
                    break;

            }
        }
    }


    private void setScript(String script) {
        if (scriptFragment == null) {
            // fragment was never intialized
            Util.showToastShort(this, "Script tab not visible??");
        } else {
            scriptFragment.setScript(script);
        }
    }



    private void createWidget(String name) {
        int mAppWidgetId;

        Intent intent = getIntent();
        Bundle extras = intent.getExtras();
        updateFilterFromFragments();
        if (extras != null) {
            mAppWidgetId = extras.getInt(
                    AppWidgetManager.EXTRA_APPWIDGET_ID,
                    AppWidgetManager.INVALID_APPWIDGET_ID);

            Context context = TodoApplication.getAppContext();

            // Store widget filter
            SharedPreferences preferences = context.getSharedPreferences("" + mAppWidgetId, MODE_PRIVATE);
            mFilter.setName(name);
            mFilter.saveInPrefs(preferences);

            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            MyAppWidgetProvider.updateAppWidget(context, appWidgetManager,
                    mAppWidgetId, name);

            Intent resultValue = new Intent(getApplicationContext(), AppWidgetService.class);
            resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId);
            setResult(RESULT_OK, resultValue);
            finish();
        }
    }

    private void applyFilter() {
        Intent data = createFilterIntent();
        startActivity(data);
        finish();
    }

    private void askWidgetName() {
        String name;
        AlertDialog.Builder alert = new AlertDialog.Builder(this);

        alert.setTitle("Create widget");
        alert.setMessage("Widget title");
        updateFilterFromFragments();
        name = mFilter.getProposedName();

// Set an EditText view to get user input
        final EditText input = new EditText(this);
        alert.setView(input);
        input.setText(name);

        alert.setPositiveButton("Ok", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int whichButton) {
                        String value = input.getText().toString();
                        if (value.equals("")) {
                            Util.showToastShort(getApplicationContext(), R.string.widget_name_empty);
                        } else {
                            createWidget(value);
                        }
                    }
                }
        );

        alert.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface dialog, int whichButton) {
                // Canceled.
            }
        });

        alert.show();

    }


    /**
     * A simple pager adapter that represents 5 ScreenSlidePageFragment objects, in
     * sequence.
     */
    private class ScreenSlidePagerAdapter extends FragmentStatePagerAdapter {
        private final ArrayList<Fragment> fragments;
        private FragmentManager fm;

        public ScreenSlidePagerAdapter(FragmentManager fm) {
            super(fm);
            this.fm=fm;
            fragments = new ArrayList<>();
        }

        public void add(Fragment frag) {
            fragments.add(frag);
        }

        @Override
        public CharSequence	getPageTitle(int position) {
            Fragment f = fragments.get(position);
            String type = f.getArguments().getString(TAB_TYPE,"unknown");
            switch ( type) {
                case PROJECT_TAB:
                    return m_app.getTagTerm();
                case CONTEXT_TAB:
                    return m_app.getListTerm();
                default:
                    return type;
            }
        }

        @Override
        public Fragment getItem(int position) {
            return fragments.get(position);
        }

        @Override
        public int getCount() {
            return fragments.size();
        }

        public ArrayList<Fragment> getFragments() {
            return fragments;
        }
    }
}

File: app/src/main/java/nl/mpcjanssen/simpletask/FilterActivity.kt
package nl.mpcjanssen.simpletask

import android.app.Activity
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import com.google.android.material.tabs.TabLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import nl.mpcjanssen.simpletask.remote.FileDialog
import nl.mpcjanssen.simpletask.remote.FileStore
import nl.mpcjanssen.simpletask.task.Priority
import nl.mpcjanssen.simpletask.task.TodoList
import nl.mpcjanssen.simpletask.util.*
import java.io.File
import java.io.IOException
import java.util.*

class FilterActivity : ThemedNoActionBarActivity() {

    internal var asWidgetConfigure = false
    internal var asWidgetReConfigure = false
    internal lateinit var mFilter: Query

    internal lateinit var m_app: TodoApplication
    val prefs = TodoApplication.config.prefs

    private var pager: ViewPager? = null
    private var m_menu: Menu? = null
    private var pagerAdapter: ScreenSlidePagerAdapter? = null
    private var scriptFragment: FilterScriptFragment? = null
    private var m_page = 0

    override fun onBackPressed() {
        if (!asWidgetConfigure && !asWidgetReConfigure) {
            applyFilter()
        }
        super.onBackPressed()
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        Log.i(TAG, "Called with intent: " + intent.toString())
        m_app = application as TodoApplication

        setContentView(R.layout.filter)
        val toolbar = findViewById<Toolbar>(R.id.toolbar_edit_filter)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeAsUpIndicator(R.drawable.ic_close_white_24dp)

        val intent = intent
        val environment: String = intent.action?.let {
            asWidgetConfigure = it == AppWidgetManager.ACTION_APPWIDGET_CONFIGURE
            val id = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, 0)
            "widget" + id.toString()
        } ?: "mainui"


        val context = applicationContext

        if (!asWidgetConfigure) {
            mFilter = Query(intent, luaModule = environment)
        } else if (intent.getBooleanExtra(Constants.EXTRA_WIDGET_RECONFIGURE, false)) {
            asWidgetReConfigure = true
            asWidgetConfigure = false
            setTitle(R.string.config_widget)
            val prefsName = intent.getIntExtra(Constants.EXTRA_WIDGET_ID, -1).toString()
            val preferences = context.getSharedPreferences(prefsName , Context.MODE_PRIVATE)
            mFilter = Query(preferences, luaModule = environment)
        } else {
            setTitle(R.string.create_widget)
            mFilter = Query(prefs, luaModule = environment)
        }
        pagerAdapter = ScreenSlidePagerAdapter(supportFragmentManager)

        val contextTab = FilterListFragment()
        contextTab.arguments = Bundle().apply {
            val contexts = alfaSort(TodoApplication.todoList.contexts, TodoApplication.config.sortCaseSensitive, "-")
            putStringArrayList(FILTER_ITEMS, contexts)
            putStringArrayList(INITIAL_SELECTED_ITEMS, mFilter.contexts)
            putBoolean(INITIAL_NOT, mFilter.contextsNot)
            putString(TAB_TYPE, CONTEXT_TAB)
        }
        pagerAdapter!!.add(contextTab)

        val projectTab = FilterListFragment()
        projectTab.arguments = Bundle().apply {
            val projects = alfaSort(TodoApplication.todoList.projects, TodoApplication.config.sortCaseSensitive, "-")
            putStringArrayList(FILTER_ITEMS, projects)
            putStringArrayList(INITIAL_SELECTED_ITEMS, mFilter.projects)
            putBoolean(INITIAL_NOT, mFilter.projectsNot)
            putString(TAB_TYPE, PROJECT_TAB)
        }
        pagerAdapter!!.add(projectTab)

        val prioTab = FilterListFragment()
        prioTab.arguments = Bundle().apply {
            putStringArrayList(FILTER_ITEMS, Priority.inCode(TodoApplication.todoList.priorities))
            putStringArrayList(INITIAL_SELECTED_ITEMS, Priority.inCode(mFilter.priorities))
            putBoolean(INITIAL_NOT, mFilter.prioritiesNot)
            putString(TAB_TYPE, PRIO_TAB)
        }
        pagerAdapter!!.add(prioTab)

        val otherTab = FilterOtherFragment()
        otherTab.arguments = Bundle().apply {
            putBoolean(Query.INTENT_HIDE_COMPLETED_FILTER, mFilter.hideCompleted)
            putBoolean(Query.INTENT_HIDE_FUTURE_FILTER, mFilter.hideFuture)
            putBoolean(Query.INTENT_HIDE_LISTS_FILTER, mFilter.hideLists)
            putBoolean(Query.INTENT_HIDE_TAGS_FILTER, mFilter.hideTags)
            putBoolean(Query.INTENT_HIDE_CREATE_DATE_FILTER, mFilter.hideCreateDate)
            putBoolean(Query.INTENT_HIDE_HIDDEN_FILTER, mFilter.hideHidden)
            putBoolean(Query.INTENT_CREATE_AS_THRESHOLD, mFilter.createIsThreshold)
            putString(TAB_TYPE, OTHER_TAB)
        }
        pagerAdapter!!.add(otherTab)

        // Fill arguments for fragment
        val sortTab = FilterSortFragment()
        sortTab.arguments = Bundle().apply {
            putStringArrayList(FILTER_ITEMS, mFilter.getSort(TodoApplication.config.defaultSorts))
            putString(TAB_TYPE, SORT_TAB)
        }
        pagerAdapter!!.add(sortTab)


        val scriptTab = FilterScriptFragment()
        scriptFragment = scriptTab
        scriptTab.arguments = Bundle().apply {
            putString(Query.INTENT_LUA_MODULE, environment)
            putBoolean(Query.INTENT_USE_SCRIPT_FILTER, mFilter.useScript)
            putString(Query.INTENT_SCRIPT_FILTER, mFilter.script)
            putString(Query.INTENT_SCRIPT_TEST_TASK_FILTER, mFilter.scriptTestTask)
            putString(TAB_TYPE, SCRIPT_TAB)
        }
        pagerAdapter!!.add(scriptTab)

        pager = findViewById<ViewPager>(R.id.pager)
        pager!!.adapter = pagerAdapter
        // Give the TabLayout the ViewPager
        val tabLayout = findViewById<TabLayout>(R.id.sliding_tabs)
        tabLayout.setupWithViewPager(pager as ViewPager)
        tabLayout.tabMode = TabLayout.MODE_SCROLLABLE
        pager?.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {
                return
            }

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                return
            }

            override fun onPageSelected(position: Int) {
                Log.i(TAG, "Page $position selected")
                m_page = position
            }
        })
        val activePage = prefs.getInt(getString(R.string.last_open_filter_tab), 0)
        if (activePage < pagerAdapter?.count ?: 0) {
            pager?.setCurrentItem(activePage, false)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        val inflater = menuInflater
        inflater.inflate(R.menu.filter, menu)
        m_menu = menu
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                finish()
                return true
            }
            R.id.menu_filter_action -> {
                when {
                    asWidgetConfigure -> askWidgetName()
                    asWidgetReConfigure -> {
                        updateWidget()
                        finish()
                    }
                    else -> applyFilter()
                }
            }
            R.id.menu_filter_load_script -> openScript { contents ->
                runOnMainThread(
                        Runnable { setScript(contents) })
            }
        }
        return true
    }

    private fun openScript(file_read: (String) -> Unit) {
            val dialog = FileDialog()
        dialog.addFileListener(object : FileDialog.FileSelectedListener {
                override fun fileSelected(file: File) {
                    Thread(Runnable {
                        try {
                            FileStore.readFile(file, file_read)
                        } catch (e: IOException) {
                            showToastShort(this@FilterActivity, "Failed to load script.")
                            e.printStackTrace()
                        }
                    }).start()
                }
            })
            dialog.createFileDialog(this@FilterActivity, FileStore, TodoApplication.config.todoFile.parentFile, txtOnly = false)
    }

    private fun createFilterIntent(): Intent {
        val target = Intent(this, Simpletask::class.java)
        target.action = Constants.INTENT_START_FILTER
        updateFilterFromFragments()
        mFilter.saveInIntent(target)

        target.putExtra("name", mFilter.proposedName)
        return target
    }

    private fun updateFilterFromFragments() {
        for (f in pagerAdapter!!.fragments) {
            when (f.arguments?.getString(TAB_TYPE, "")?: "") {
                "" -> {
                }
                OTHER_TAB -> {
                    val of = f as FilterOtherFragment
                    mFilter.hideCompleted = of.hideCompleted
                    mFilter.hideFuture = of.hideFuture
                    mFilter.hideLists = of.hideLists
                    mFilter.hideTags = of.hideTags
                    mFilter.hideCreateDate = of.hideCreateDate
                    mFilter.hideHidden = of.hideHidden
                    mFilter.createIsThreshold = of.createAsThreshold
                }
                CONTEXT_TAB -> {
                    val lf = f as FilterListFragment
                    mFilter.contexts = lf.getSelectedItems()
                    mFilter.contextsNot = lf.getNot()
                }
                PROJECT_TAB -> {
                    val pf = f as FilterListFragment
                    mFilter.projects = pf.getSelectedItems()
                    mFilter.projectsNot = pf.getNot()
                }
                PRIO_TAB -> {
                    val prf = f as FilterListFragment
                    mFilter.priorities = Priority.toPriority(prf.getSelectedItems())
                    mFilter.prioritiesNot = prf.getNot()
                }
                SORT_TAB -> {
                    val sf = f as FilterSortFragment
                    mFilter.setSort(sf.selectedItem)
                }
                SCRIPT_TAB -> {
                    val scrf = f as FilterScriptFragment
                    mFilter.useScript = scrf.useScript
                    mFilter.script = scrf.script
                    mFilter.scriptTestTask = scrf.testTask
                }
            }
        }
    }

    private fun setScript(script: String?) {
        if (scriptFragment == null) {
            // fragment was never intialized
            showToastShort(this, "Script tab not visible??")
        } else {
            script?.let { scriptFragment!!.script = script }
        }
    }

    private fun updateWidget() {
        updateFilterFromFragments()
        val widgetId = intent.getIntExtra(Constants.EXTRA_WIDGET_ID, 0)
        Log.i(TAG, "Saving settings for widget $widgetId")
        val preferences = applicationContext.getSharedPreferences("" + widgetId, Context.MODE_PRIVATE)
        mFilter.saveInPrefs(preferences)
        broadcastRefreshWidgets(m_app.localBroadCastManager)
    }

    private fun createWidget(name: String) {
        val mAppWidgetId: Int

        val intent = intent
        val extras = intent.extras
        updateFilterFromFragments()
        if (extras != null) {
            mAppWidgetId = extras.getInt(
                    AppWidgetManager.EXTRA_APPWIDGET_ID,
                    AppWidgetManager.INVALID_APPWIDGET_ID)

            val context = applicationContext

            // Store widget applyFilter
            val preferences = context.getSharedPreferences("" + mAppWidgetId, Context.MODE_PRIVATE)
            val namedFilter = NamedQuery(name, mFilter)
            namedFilter.saveInPrefs(preferences)

            val appWidgetManager = AppWidgetManager.getInstance(context)
            MyAppWidgetProvider.updateAppWidget(context, appWidgetManager,
                    mAppWidgetId, name)

            val resultValue = Intent(applicationContext, AppWidgetService::class.java)
            resultValue.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, mAppWidgetId)
            setResult(Activity.RESULT_OK, resultValue)
            finish()
        }
    }

    private fun applyFilter() {
        val data = createFilterIntent()
        startActivity(data)
        finish()
    }

    private fun askWidgetName() {
        val name: String
        val alert = AlertDialog.Builder(this)

        alert.setTitle("Create widget")
        alert.setMessage("Widget title")
        updateFilterFromFragments()
        name = mFilter.proposedName

        // Set an EditText view to get user input
        val input = EditText(this)
        alert.setView(input)
        input.setText(name)

        alert.setPositiveButton("Ok") { _, _ ->
            val value = input.text.toString()
            if (value == "") {
                showToastShort(applicationContext, R.string.widget_name_empty)
            } else {
                createWidget(value)
            }
        }

        alert.setNegativeButton("Cancel") { _, _ -> }

        alert.show()

    }

    override fun onDestroy() {
        super.onDestroy()
        prefs.edit().putInt(getString(R.string.last_open_filter_tab), m_page).apply()
        pager?.clearOnPageChangeListeners()
    }

    /**
     * A simple pager adapter that represents 5 ScreenSlidePageFragment objects, in
     * sequence.
     */
    private inner class ScreenSlidePagerAdapter(fm: FragmentManager) : FragmentStatePagerAdapter(fm) {
        val fragments: ArrayList<Fragment>

        init {
            fragments = ArrayList<Fragment>()
        }

        fun add(frag: Fragment) {
            fragments.add(frag)
        }

        override fun getPageTitle(position: Int): CharSequence {
            val f = fragments[position]
            val type = f.arguments?.getString(TAB_TYPE, "unknown") ?:"unknown"
            when (type) {
                PROJECT_TAB -> return TodoApplication.config.tagTerm
                CONTEXT_TAB -> return TodoApplication.config.listTerm
                else -> return type
            }
        }

        override fun getItem(position: Int): Fragment {
            return fragments[position]
        }

        override fun getCount(): Int {
            return fragments.size
        }

    }

    companion object {

        val TAG = "FilterActivity"
        val TAB_TYPE = "type"
        val CONTEXT_TAB = "context"
        val PROJECT_TAB = "project"
        val PRIO_TAB = getString(R.string.filter_tab_header_prio)
        val OTHER_TAB = getString(R.string.filter_tab_header_other)
        val SORT_TAB = getString(R.string.filter_tab_header_sort)
        val SCRIPT_TAB = getString(R.string.filter_tab_header_script)

        // Constants for saving state
        val FILTER_ITEMS = "items"
        val INITIAL_SELECTED_ITEMS = "initialSelectedItems"
        val INITIAL_NOT = "initialNot"
    }
}

